### PR TITLE
Introduce Hubble HTTP v2 metrics and dashboards

### PR DESCRIPTION
**Pull request by chancez**
This is based on #21079 and cherry-picks the changes from #21124 since the dashboards generally require both of those changes. 

http v2 metrics are based on L7 response flows rather than both request
and response flows. There was previously some overlap between
http_requests_total and http_responses_total, with the latter having the
status codes. In practice, we should only need one of these, but it
needs the request status code. To achieve this, we now only use the
response flows to count requests, and we invert the source/destination
of the response flows to get the source/destination from the perspective
of the request, rather than the response.

We already have http_request_duration_seconds, which is based on
responses already, but this also leads to the source/destination labels
being "backwards". So like we do with the new http_requests_total
metric, we invert the source and destination.

In addition to consolidating, this is consistent with how http metrics
are exported in the wider ecosystem.

For backwards compatibility, we keep the existing behavior by default,
and expose the new v2 metrics with the same metric names as before using
a feature flag in the metrics context. HTTP v2 metrics can be enabled by
specifying `httpV2` in the `hubble.metric.enabled` helm value:

```
hubble:
  metrics:
    enabled: [ "httpV2" ]
```

Additionally, this PR adds a dashboard which uses the v2 metrics. The dashboard is included as part of the helm chart so that it can be automatically loaded into Grafana when using Grafana installed via the [kube-prometheus-stack helm chart](https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-prometheus-stack). 

For example, if `kube-prometheus-stack` is installed in the `monitoring` namespace:
```
hubble:
  metrics:
    dashboards:
      enabled: true
      namespace: monitoring
```

Here's a screenshot of the dashboard:

Screen Shot 2022-09-01 at 2 04 31 PM

I think it would be nice to have request sizes like Istio, but for now this is a good start!

I also did some refactoring in the metrics context options code to make the logic of inverting the source/destination (which is used by http v2 metrics) simpler.

Release note:
```release-note
Introduce Hubble HTTP v2 metrics and dashboards
```

### CODE REVIEW

**Review by squeed — APPROVED** [review]

File: install/kubernetes/cilium/templates/hubble/dashboards-configmap.yaml [hunk 1]
```diff
@@ -0,0 +1,29 @@
+{{- if .Values.hubble.metrics.dashboards.enabled }}
+{{- $files := .Files.Glob "files/hubble/dashboards/*.json" }}
+{{- if $files }}
+apiVersion: v1
+kind: ConfigMapList
+items:
+{{- range $path, $fileContents := $files }}
+{{- $dashboardName := regexReplaceAll "(^.*/)(.*)\\.json$" $path "${2}" }}
```
Comment (kaworu) [inline install/kubernetes/cilium/templates/hubble/dashboards-configmap.yaml:8]: nit: strange to see the start of string anchor inside the first capture group, also why two capture groups if we only extract one?
```suggestion
{{- $dashboardName := regexReplaceAll "^.*/(.*)\\.json$" $path "${1}" }}
```

File: pkg/hubble/metrics/api/context.go [hunk 2]
```diff
@@ -352,6 +349,13 @@ func handleReservedIdentityLabels(lbls []string) string {
 	return ""
 }
 
+func destinationNamespaceContext(flow *pb.Flow) (context string) {
+	if flow.GetDestination() != nil {
+		context = flow.GetDestination().Namespace
+	}
+	return
+}
```
Comment (kaworu) [inline pkg/hubble/metrics/api/context.go]: this func only moved right? Any reason behind re-ordering?

File: pkg/hubble/metrics/api/context.go [hunk 3]
```diff
@@ -418,71 +309,101 @@ func getK8sAppFromLabels(labels []string) string {
...
+		if destinationLabel != "" {
+			break
 		}
-		labels = append(labels, context)
 	}
 
+	if invert {
+		sourceLabel, destinationLabel = destinationLabel, sourceLabel
+	}
+	if len(o.Source) != 0 {
```
Comment (kaworu) [inline pkg/hubble/metrics/api/context.go:360]: Shouldn't the condition rather be

```suggestion
	if sourceLabel != "" {
```

File: pkg/hubble/metrics/api/context.go [hunk 4]
```diff
@@ -214,17 +214,14 @@ func (ls labelsSet) HasLabel(label string) bool {
...
-			b.WriteString(label)
-			if i < len(ls)-1 {
-				b.WriteString(",")
-			}
+			labels = append(labels, label)
 		}
 	}
-	return b.String()
+	return strings.Join(labels, ",")
 }
```
Comment (kaworu) [inline pkg/hubble/metrics/api/context.go]: Was the func broken before this patch? Asking for clarification because the commit message is "Fix labelSet.String()" while I don't see the bug. If not, why the change?

**Review by kaworu — COMMENTED** [review]
Awesome @chancez! Couple of comments / questions.

File: install/kubernetes/cilium/templates/hubble/dashboards-configmap.yaml (hunk 1 above)
Comment (chancez) [reply install/kubernetes/cilium/templates/hubble/dashboards-configmap.yaml:8]: To be honest, I copied this from https://github.com/prometheus-community/helm-charts/blob/main/charts/kube-prometheus-stack/templates/grafana/configmap-dashboards.yaml and I didn't really look too deeply at this.

File: pkg/hubble/metrics/api/context.go (hunk 2 above)
Comment (chancez) [reply pkg/hubble/metrics/api/context.go]: I think this only moved in an intermediate commit, it was just weird that this wasn't grouped with the other destination* options. In a later commit I refactored all this so we don't even have this function anymore. https://github.com/cilium/cilium/pull/21181/commits/bd7b28723bc995c87411ef9b2ba93a03c109d01d

File: pkg/hubble/metrics/api/context.go (hunk 4 above)
Comment (chancez) [reply pkg/hubble/metrics/api/context.go]: It was improperly formatting before, I had noticed issues in the implementation when I was reading the logs at startup. This was incorrectly implemented in the labelsContext PR. All my fault 😭

File: pkg/hubble/metrics/api/context.go (hunk 3 above)
Comment (chancez) [reply pkg/hubble/metrics/api/context.go:360]: sourceLabel will potentially still be `""` when no context options match, even if they're supplied, but we don't want to include the label unless the context option is supplied because we will have a mismatched number of label values and label names, so we have to check `o.Source` (which is only non-empty if the sourceContext is supplied).

File: pkg/hubble/metrics/api/context.go (hunk 4 above)
Comment (kaworu) [reply pkg/hubble/metrics/api/context.go]: If this is a fix it would be great to add a small unit test, not critical though.

**Review by kaworu — APPROVED** [review]
Thanks @chancez!

File: Documentation/observability/metrics.rst [hunk 5]
```diff
@@ -650,6 +653,9 @@ This metric supports :ref:`Context Options<hubble_context_options>`.
 ``http``
 ~~~~~~~~
 
+Deprecated, use ``httpV2`` instead.
```
Comment (qmonnet) [inline Documentation/observability/metrics.rst:656]: Should there be a note in Documentation/operations/upgrade.rst about this deprecation? Is this supposed to be removed in a future version?

File: Documentation/observability/metrics.rst [hunk 6]
```diff
@@ -671,6 +677,42 @@ Options
...
+``http_request_duration_seconds`` ``method``, ``reporter``                            Disabled   Quantiles of HTTP request duration in seconds
+================================= =================================================== ========== ==============================================
+
+Labels
+""""""
+
+- ``method`` is the HTTP method of the request/response.
+- ``protocol`` is the HTTP protocol of the request, (For example: ``HTTP/1.1``, ``HTTP/2``).
+- ``status`` is the HTTP status code of the response.
+- ``reporter`` identifies the origin of the request/response. It is set to ``client`` if it originated from the client, ``server`` if it originated from the server, or ``unknown`` if it's origin isn't known.
```
Comment (qmonnet) [inline Documentation/observability/metrics.rst]:
```suggestion
- ``reporter`` identifies the origin of the request/response. It is set to ``client`` if it originated from the client, ``server`` if it originated from the server, or ``unknown`` if its origin is unknown.
- ```

File: Documentation/observability/metrics.rst (hunk 5 above)
Comment (kaworu) [reply Documentation/observability/metrics.rst:656]: Good catch, added a deprecation notice in `upgrade.rst`, removal is not planned for now.

File: Documentation/observability/metrics.rst (hunk 6 above)
Comment (kaworu) [reply Documentation/observability/metrics.rst]: Thanks, also fixed in `http`

File: Documentation/observability/metrics.rst [hunk 7]
```diff
@@ -664,7 +670,43 @@ Labels
...
+related labels now are from the perspective of the request. In the ``http``
+metrics, the source/destination were swapped, because the metric uses the
+response flow data, where the source/destination are swapped, but in ``httpV2``
+we correctly account for this.
+
+================================= =================================================== ========== ==============================================
+Name                              Labels                                              Default     Description
+================================= =================================================== ========== ==============================================
+``http_requests_total``           ``method``, ``protocol``, ``status``, ``reporter``  Disabled   Count of HTTP requests
+``http_request_duration_seconds`` ``method``, ``reporter``                            Disabled   Quantiles of HTTP request duration in seconds
```
Comment (ghost) [inline Documentation/observability/metrics.rst]: I know it's just copied over, so nothing blocking, but imo "quantiles" suggests it's a summary, not a histogram. I would rather write "Histogram of HTTP request duration in seconds" (and same for metrics help).

File: Documentation/observability/metrics.rst [hunk 8]
```diff
@@ -664,7 +670,43 @@ Labels
...
+Options
+"""""""
+
+This metric supports :ref:`Context Options<hubble_context_options>`.
+
+``httpV2``
+~~~~~~~~~~
+
+``httpV2`` is an updated version of the existing ``http`` metrics.
+These metrics should not be enabled at the same time as ``http``.
```
Comment (ghost) [inline Documentation/observability/metrics.rst]: From what I see, if both are enabled then only `httpV2` metrics are exposed? I guess it's the best option (as `http_request_duration_seconds` has the same labels), but would be nice to document that, also in the upgrade notes - the migration path is a little tricky if you can't have both old and new at the same time.

**Review by ghost — COMMENTED** [review]
nothing blocking, just left some comments about docs :)

**Review by qmonnet — APPROVED** [review]
Thanks!

File: Documentation/observability/metrics.rst (hunk 7 above)
Comment (kaworu) [reply Documentation/observability/metrics.rst]: No problem Anna, the matching `http` section be updated as well correct?

Comment (ghost) [reply Documentation/observability/metrics.rst]: Ideally, I think so.

File: Documentation/observability/metrics.rst (hunk 8 above)
Comment (kaworu) [reply Documentation/observability/metrics.rst]: My understanding is that it is not possible to enable both `http` and `httpV2`, the second commit in the stack ("hubble: Ensure http and httpV2 metrics cannot both be enabled") implement the conflict detection.

Admittedly, the doc could be clearer about the incompatibility (i.e. state "_can not_ be enabled at the same time […]" rather than "_should not_ be enabled at the same time […]". Does that make sense?

Comment (ghost) [reply Documentation/observability/metrics.rst]: Yeah, makes sense to me. My worry here was mainly about the migration path - if we say both old and new "should not" be enabled at the same time, it might be tempting to enable both for a short period, then migrate dashboards, alerts, etc, then disable old. But it's not possible.

**Comment by kaworu** [comment]
@lambdanis addressed both comments, please take another look :pray:

**Review by ghost — APPROVED** [review]

**Comment by kaworu** [comment]
/test-1.16-4.9

File: Documentation/observability/metrics.rst [hunk 9]
```diff
@@ -664,7 +670,43 @@ Labels
...
+data.
+
+Additionally, the ``http_request_duration_seconds`` metric source/destination
+related labels now are from the perspective of the request. In the ``http``
+metrics, the source/destination were swapped, because the metric uses the
+response flow data, where the source/destination are swapped, but in ``httpV2``
+we correctly account for this.
+
+================================= =================================================== ========== ==============================================
+Name                              Labels                                              Default     Description
```
Comment (christarazi) [inline Documentation/observability/metrics.rst]: Very nit: The `Description` column header has an extra space prepended.

File: pkg/hubble/metrics/api/context.go [hunk 10]
```diff
@@ -214,17 +214,14 @@ func (ls labelsSet) HasLabel(label string) bool {
...
 
 func (ls labelsSet) String() string {
-	var b strings.Builder
+	var labels []string
```
Comment (christarazi) [inline pkg/hubble/metrics/api/context.go]: The commit says "Fix", but what is it fixing and why should be described in the commit msg.

**Comment by gandro** [comment]
Removing the automatic ready-to-merge label. If Chris' nits are not to be addressed, feel free to add again.

File: pkg/hubble/metrics/api/context.go (hunk 10 above)
Comment (kaworu) [reply pkg/hubble/metrics/api/context.go]: Hi @christarazi good point, I refactored the patch to add unit tests (including a test case for the bug fixed by the commit), refactored the implementation to avoid a temporary slice, and added more info in the commit message. Please take another look :pray:

File: Documentation/observability/metrics.rst (hunk 9 above)
Comment (kaworu) [reply Documentation/observability/metrics.rst]: Good catch!

**Review by christarazi — APPROVED** [review]

**Comment by kaworu** [comment]
/test-1.25-net-next

**Comment by stevo-f3** [comment]
@chancez please consider putting this config map additionally behind `hubble.enabled` flag

When creating preflight manifest like documented in https://docs.cilium.io/en/latest/operations/upgrade/#running-pre-flight-check-required it should be enough to disable hubble, and not to require to also disable the dashboard.

**Comment by chancez** [comment]
PR: https://github.com/cilium/cilium/pull/22161
